lbrynet/core/LBRYWallet.py

```diff
@@ -355,34 +366,68 @@ def _build_response(m, result):
 
         _check_result_fields(result)
 
-        try:
-            metadata = Metadata(json.loads(result['value']))
-        except (ValueError, TypeError):
-            return Failure(InvalidStreamInfoError(name))
+        txid = result['txid']
 
-        d = self._save_name_metadata(name, str(result['txid']), metadata['sources']['lbry_sd_hash'])
-        d.addCallback(lambda _: log.info("lbry://%s complies with %s" % (name, metadata.meta_version)))
-        d.addCallback(lambda _: _build_response(metadata, result))
+        if force_good_metadata:
+            try:
+                metadata = Metadata(json.loads(result['value']))
+            except (ValueError, TypeError):
+                return Failure(InvalidStreamInfoError(name))
+            sd_hash = metadata['sources']['lbry_sd_hash']
+            d = self._save_name_metadata(name, txid, sd_hash)
+            d.addCallback(lambda _: self.get_claimid(name, txid))
+        else:
+            metadata = result['value']
+            d = self.get_claimid(name, txid)
+        d.addCallback(lambda claim_id: _build_response(metadata, result, claim_id))
         return d
 
-    def get_claim_info(self, name):
-        d = self._get_value_for_name(name)
-        d.addCallback(lambda r: self._get_claim_info(r, name))
-        return d
+    def get_claimid(self, name, txid):
+        def _get_id_for_return(claim_id):
+            if claim_id:
+                return defer.succeed(claim_id)
+            else:
+                d = self.get_claims_from_tx(txid)
+                d.addCallback(lambda claims: next(c['claimId'] for c in claims if c['name'] == name))
+                d.addCallback(lambda cid: self._update_claimid(cid, name, txid))
+                return d
 
+        d = self._get_claimid_for_tx(name, txid)
+        d.addCallback(_get_id_for_return)
+        return d
 
-    def claim_name(self, name, bid, m):
+    def get_claim_info(self, name, force_good_metadata=True, is_mine=False):
```

Comment:
same complaint about boolean arguments here.  Functions should be small and *do one thing*. 

The requiring that the claim be your own can moved outside of this function and happen by the caller:

```
d = get_claim_info(...)
d.addCallback(ensure_my_claim)
```

For `force_good_metadata`, I need to think more about a concrete suggestion.  There probably needs to be a larger refactoring that allows `get_claim_info` to be extensible. (https://en.wikipedia.org/wiki/Open/closed_principle)